catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java

```diff
@@ -0,0 +1,161 @@
+import java.util.List;
+import java.util.zip.GZIPOutputStream;
+import javax.ws.rs.core.HttpHeaders;
+import org.apache.tika.mime.MimeTypeException;
+import org.apache.tika.mime.MimeTypes;
+import org.boon.json.JsonParserFactory;
+import org.boon.json.JsonSerializerFactory;
+import org.boon.json.ObjectMapper;
+import org.boon.json.implementation.ObjectMapperImpl;
+import org.codice.ddf.catalog.ui.query.cql.CqlQueryResponse;
+import org.codice.ddf.catalog.ui.query.cql.CqlRequest;
+import org.codice.ddf.catalog.ui.util.EndpointUtil;
+import org.eclipse.jetty.http.HttpStatus;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.Request;
+import spark.Response;
+import spark.Route;
+import spark.utils.IOUtils;
+
+public class CqlTransformHandler implements Route {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CqlTransformHandler.class);
+
+  private List<ServiceReference> queryResponseTransformers;
+
+  private BundleContext bundleContext;
+
+  private static final String GZIP = "gzip";
+
+  private ObjectMapper mapper =
+      new ObjectMapperImpl(
+          new JsonParserFactory().usePropertyOnly(),
+          new JsonSerializerFactory()
+              .includeEmpty()
+              .includeNulls()
+              .includeDefaultValues()
+              .setJsonFormatForDates(false)
+              .useAnnotations());
+
+  private EndpointUtil util;
+
+  public CqlTransformHandler(
+      List<ServiceReference> queryResponseTransformers, BundleContext bundleContext) {
+    this.queryResponseTransformers = queryResponseTransformers;
+    this.bundleContext = bundleContext;
+  }
+
+  @Override
+  public Object handle(Request request, Response response) throws Exception {
+    String transformerId = request.params(":transformerId");
+    String body = util.safeGetBody(request);
+    CqlRequest cqlRequest = mapper.readValue(body, CqlRequest.class);
+    CqlQueryResponse cqlQueryResponse = util.executeCqlQuery(cqlRequest);
+
+    LOGGER.trace("Finding transformer to transform query response.");
+
+    ServiceReference<QueryResponseTransformer> queryResponseTransformer =
+        queryResponseTransformers
+            .stream()
+            .filter(transformer -> transformer.getProperty("id").equals(transformerId))
+            .findFirst()
+            .orElse(null);
+
+    if (queryResponseTransformer == null) {
+      LOGGER.debug("Could not find transformer with id: {}", transformerId);
+      response.status(HttpStatus.NOT_FOUND_404);
+      return mapper.toJson(ImmutableMap.of("message", "Service not found"));
+    }
+
+    attachFileToResponse(request, response, queryResponseTransformer, cqlQueryResponse);
+
+    return "";
+  }
+
+  private void setHttpHeaders(Request request, Response response, BinaryContent content)
+      throws MimeTypeException {
+    String mimeType = content.getMimeTypeValue();
```

Comment:
can this ever return null?

Reply:
yes. like above, i ran into that issue with testing so i'll also add a check into that